File: algorithms/sjf/java/sjf-sort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.stream.Collectors;

class Process {
    int pid, at, bt, ct;

    public Process(int pid, int at, int bt) {
        this.pid = pid;
        this.at = at;
        this.bt = bt;
        this.ct = 0;
    }
}

class sjf {
    public static void main(String[] args) {
        ArrayList<Process> proc = new ArrayList<Process>(
                Arrays.asList(new Process(0, 13, 5), new Process(1, 0, 5), new Process(2, 0, 2), new Process(3, 2, 4)));
        float ta = 0, w = 0;

        Collections.sort(proc, (a, b) -> (a.bt < b.bt) ? -1 : 1);
        Collections.sort(proc, (a, b) -> (a.at < b.at) ? -1 : 1);
        proc.get(0).ct = proc.get(0).at + proc.get(0).bt;

        int maxBT = Collections.max(proc, (a, b) -> (a.bt < b.bt) ? -1 : 1).bt;
        int end = 1, time = proc.get(0).ct;
        while (end < proc.size()) {
            int idx = -1, minBT = maxBT;
            for (int i = 1; i < proc.size(); i++)
                if ((proc.get(i).ct == 0) && (proc.get(i).bt <= minBT) && (proc.get(i).at <= time)) {
                    if ((idx != -1) && proc.get(i).bt == minBT && proc.get(i).at >= proc.get(idx).at)
                        continue;
                    minBT = proc.get(i).bt;
                    idx = i;
                }
            if (idx == -1) {
                final int curr = time;
                var prc = Collections.min(proc.stream().filter(p -> (p.at > curr)).collect(Collectors.toList()),
                        (a, b) -> (a.at < b.at) ? 1 : -1);
                time = prc.at;
                idx = proc.indexOf(prc);
            }
            time += proc.get(idx).bt;
            proc.get(idx).ct = time;
            end++;
            ta += time - proc.get(idx).at;
            w += time - (proc.get(idx).at + proc.get(idx).bt);
        }

        Collections.sort(proc, (a, b) -> (a.pid < b.pid) ? -1 : 1);

        System.out.println("PID\tAT\tBT\tCT\tTAT\tWT");
        proc.forEach(p -> System.out.println(p.pid + "\t" + p.at + '\t' + p.bt + '\t' + p.ct + '\t' + (p.ct - p.at)
                + '\t' + (p.ct - (p.at + p.bt))));
        System.out.println("Avg. TAT: " + ta / proc.size());
        System.out.println("Avg. WT: " + w / proc.size());
    }
}

File: algorithms/sjf/cpp/stl/sjf-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct Process {
	unsigned pid;
	unsigned at;
	unsigned bt;
	unsigned ct;
};

// ------------------------------------------------------
// SJF Algo Implementation

void findCompletionTime(std::vector<Process> &proc) {
	unsigned count = proc.size();

	proc[0].ct = proc[0].at + proc[0].bt;

	unsigned fin = 1, time = proc[0].ct;
	unsigned maxBT = std::max_element(
			proc.begin(),
			proc.end(),
			[] (const Process &a, const Process &b) -> bool {
			return a.bt < b.bt;
			})->bt;

	while (fin < count) {
		int idx = -1;
		unsigned min = maxBT;
		for (int i = 1; i < count; i++) {
			if (!proc[i].ct && proc[i].bt <= min && proc[i].at <= time) {
				if (
					idx != -1 && proc[i].bt == min &&
					proc[i].at >= proc[idx].at
				   )
					continue;
				min = proc[i].bt;
				idx = i;
			}
		}
		if (idx == -1)
			for (int i = 0; i < count; i++)
				if (proc[i].at > time) {
					time = proc[i].at;
					idx = i;
					break;
				}
		time += proc[idx].bt;
		proc[idx].ct = time;
		fin++;
	}
}

void findAvgTime(std::vector<Process> &proc) {
	std::sort(
			proc.begin(),
			proc.end(),
			[](const Process &a, const Process &b) {
			return a.bt < b.bt;
			});
	std::sort(
			proc.begin(),
			proc.end(),
			[](const Process &a, const Process &b) {
			return a.at < b.at;
			});
	findCompletionTime(proc);
	std::sort(
			proc.begin(),
			proc.end(),
			[](const Process &a, const Process &b) {
			return a.pid < b.pid;
			});

	float ta{0}, w{0};

	std::cout << "PID\tAT\tBT\tCT\tTAT\tWT\n";
	for (const auto &p: proc) {
		unsigned pTA = p.ct - p.at;
		unsigned pW = pTA - p.bt;
		std::cout << p.pid << '\t' << p.at << '\t'
			<< p.bt << '\t' << p.ct << '\t'
			<< pTA << '\t' << pW << '\n';
		ta += pTA;
		w += pW;
	}

	ta /= proc.size();
	w /= proc.size();

	std::cout << std::endl;
	std::cout << "Average Waiting Time: " << w << '\n' 
		<< "Average Turn-Around Time: " << ta << '\n';
}

int main() {
	std::vector<Process> proc = {
		{0, 13, 2, 0},
		{1, 5, 1, 0},
		{2, 2, 3, 0},
		{3, 1, 1, 0},
		{4, 0, 4, 0}
	};

	findAvgTime(proc);

	std::cout << '\n';
	return 0;
}
